_posts/2020-04-27-mocking.adoc

```diff
@@ -0,0 +1,375 @@
+
+=== Using @InjectMock
+
+Hopefully the previous section convinced you of the merits of `QuarkusMock` over the old approach. You might also be wondering
+however if there is a way to reduce boilerplate code even further and provide tighter integration with Mockito. That is
+where `@InjectMock` comes in handy.
+
+To demonstrate `@InjectMock` let's rewrite the `MockTestCase` from the previous section.
+
+First of all, we need to add the following dependency:
+
+[source,xml]
+----
+<dependency>
+    <groupId>io.quarkus</groupId>
+    <artifactId>quarkus-junit5-mockito</artifactId>
+</dependency>
+----
+
+Now we can rewrite the `MockTestCase` like so:
+
+[source,java]
+----
+@QuarkusTest
+public class MockTestCase {
+
+    @InjectMock
+    MockableBean1 mockableBean1; // <1>
+
+    @InjectMock
+    MockableBean2 mockableBean2;
+
+    @BeforeEach
+    public void setup() {
+        Mockito.when(mockableBean1.greet("Stuart")).thenReturn("A mock for Stuart"); // <2>
+    }
+
+    @Test
+    public void firstTest() {
+        Assertions.assertEquals("A mock for Stuart", mockableBean1.greet("Stuart"));
+        Assertions.assertEquals(null, mockableBean2.greet("Stuart"));
+    }
+
+    @Test
+    public void secondTest() {
+        Mockito.when(mockableBean2.greet("Stuart")).thenReturn("Bonjour Stuart"); // <3>
+        Assertions.assertEquals("A mock for Stuart", mockableBean1.greet("Stuart"));
+        Assertions.assertEquals("Bonjour Stuart", mockableBean2.greet("Stuart"));
+    }
+
+    @ApplicationScoped
+    public static class MockableBean1 {
+
+        public String greet(String name) {
+            return "Hello " + name;
+        }
+    }
+
+    @ApplicationScoped
+    public static class MockableBean2 {
+
+        public String greet(String name) {
+            return "Hello " + name;
+        }
+    }
+}
+----
+<1> `@InjectMock` results in a mock being created and being available in *all* test methods of the test class (other test classes are *not* affected by this)
+<2> The `mockableBean1` is configured here for all test methods of the class
+<3> Configure `mockableBean2` for this test only
+
+[NOTE]
+====
+Since `@InjectMock` uses `QuarkusMock` under the hood, the same limitations apply to its use.
+
+Additionally, `@InjectMock` works like an injection point for the bean, so for it to work properly when
+the target bean uses CDI qualifiers, those qualifiers also need to be added to the field.
+We will see an example of this in next section about mocking a `@RestClient` bean.
+====
+
+As a final example, we can rewrite the `OrderServiceTest` test like so:
+
+[source,java]
+----
+public class OrderServiceTest {
+
+    @Inject
+    private OrderService orderService;
+
+    @InjectMock
+    private InvoiceNotificationService invoiceNotificationService;
+
+    @BeforeAll
+    public static void setup() {
+        doNothing().when(mock).sendInvoice(any());
+        doNothing().when(mock).notifyInvoiceAlreadySent(any());
```

Comment:
I think this should refer to the `invoiceNotificationService` field, not `mock`

Reply:
Nice catch! Fixed